src/format.rs

```diff
@@ -0,0 +1,723 @@
+            Spec::WeekDayNameAbbr => todo!(),
+            Spec::WeekDayFrom1 => todo!(),
+            Spec::WeekDayFrom0 => todo!(),
+            Spec::WeekBasedYear => todo!(),
+            Spec::WeekBasedYearRem100 => todo!(),
+            Spec::WeekNumberOfWeekBasedYear => todo!(),
+            Spec::WeekNumberFromSunday => todo!(),
+            Spec::WeekNumberFromMonday => todo!(),
+            Spec::SecondsSinceEpoch => todo!(),
+            Spec::Newline => todo!(),
+            Spec::Tabulation => todo!(),
+            Spec::Percent => todo!(),
+            Spec::CombinationDateTime => todo!(),
+            Spec::CombinationDate => todo!(),
+            Spec::CombinationIso8601 => todo!(),
+            Spec::CombinationVmsDate => todo!(),
+            Spec::CombinationTime12h => todo!(),
+            Spec::CombinationHourMinute24h => todo!(),
+            Spec::CombinationTime24h => todo!(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TimeFormatter<'t, 'f> {
+    time: &'t Time,
+    format: &'f [u8],
+}
+
+impl<'t, 'f> TimeFormatter<'t, 'f> {
+    pub fn new<T: AsRef<[u8]> + ?Sized>(time: &'t Time, format: &'f T) -> Self {
+        Self {
+            time,
+            format: format.as_ref(),
+        }
+    }
+
+    pub fn fmt(&self, buf: &mut dyn Write) -> Result<(), FormatError> {
+        // Use a size limiter to avoid large writes
+        let size_limit = self.format.len().saturating_mul(1024 * 1024).max(1024);
+        let mut f = SizeLimiter::new(buf, size_limit);
+
+        let mut cursor = Cursor::new(self.format);
+
+        loop {
+            f.write_all(cursor.read_until(|&x| x == b'%'))?;
+
+            let remaining_before = cursor.remaining();
+
+            if cursor.next().is_none() {
+                break;
+            }
+
+            let flags = cursor.read_while(|&x| matches!(x, b'-' | b'_' | b'0' | b'^' | b'#'));
+
+            let mut padding = Padding::Left;
+            let mut has_left_padding = false;
+            let mut case = None;
+
+            for &flag in flags {
+                match flag {
+                    b'-' => {
+                        padding = Padding::Left;
+                        has_left_padding = true;
+                    }
+                    b'_' => padding = Padding::Spaces,
+                    b'0' => padding = Padding::Zeros,
+                    b'^' if case != Some(Case::Change) => case = Some(Case::Upper),
+                    b'#' => case = Some(Case::Change),
+                    _ => (),
+                }
+            }
+
+            let width = str::from_utf8(cursor.read_while(u8::is_ascii_digit))
+                .unwrap()
+                .parse()
+                .ok();
+
+            if let Some(
+                b"Ec" | b"EC" | b"Ex" | b"EX" | b"Ey" | b"EY" | b"Od" | b"Oe" | b"OH" | b"OI"
+                | b"Om" | b"OM" | b"OS" | b"Ou" | b"OU" | b"OV" | b"Ow" | b"OW" | b"Oy",
+            ) = cursor.remaining().get(..2)
```

Comment:
it took me a bit to figure out what this `if` is doing, but looking at the docs for `Time#strftime`, it looks like there are certain modifiers that are ignored in certain positions? Can we quote the relevant docs here in a comment to serve as a clarification?